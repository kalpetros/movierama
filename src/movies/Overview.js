import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';

export const Overview = (props) => {
  const [data, setData] = useState([]);
  const { id: id } = props;

  useEffect(() => {
    const fetchDetails = () => {
      const apiKey = 'bc50218d91157b1ba4f142ef7baaa6a0';
      const endpoint = `https://api.themoviedb.org/3/movie/${id}?api_key=${apiKey}`;

      fetch(endpoint, {
        method: 'GET',
      })
        .then((response) => {
          if (response.status !== 200) throw new Error(response.status);
          return response.json();
        })
        .then((data) => {
          setData(data);
        })
        .catch((error) => {
          console.log(error);
        });
    };

    fetchDetails();
  }, [id]);

  if (Object.keys(data).length === 0) {
    return null;
  }

  const overview = data.overview;
  const genres = data.genres.map((genre, index) => {
    return genre.name;
  });
  const video = data.video;
  const genresText = genres.join(', ');

  const producers = data.production_companies.map((company, index) => {
    return company.name;
  });

  const producersText = producers.length > 0 ? producers.join(', ') : 'N/A';

  const countries = data.production_countries.map((country, index) => {
    return country.name;
  });

  const countriesText = countries.length > 0 ? countries.join(', ') : 'N/A';

  const runtime = `${data.runtime} minutes`;

  return (
    <div className="movie__details__meta">
      <small className="font-bold text-uppercase">{genresText}</small>
      <p>{overview}</p>
      <h3>Runtime</h3>
      <small className="font-bold text-uppercase">{runtime}</small>
      <h3>Producers</h3>
      <small className="font-bold text-uppercase">{producersText}</small>
      <h3>Countries</h3>
      <small className="font-bold text-uppercase">{countriesText}</small>
    </div>
  );
};

Overview.defaultProps = {
  id: 1,
};

Overview.propTypes = {
  id: PropTypes.number.isRequired,
};
