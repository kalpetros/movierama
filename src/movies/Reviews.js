import React, { useState, useEffect } from 'react';

export const Reviews = (props) => {
  const [data, setData] = useState([]);
  const { id: id } = props;

  useEffect(() => {
    fetchReviews();
  }, []);

  const fetchReviews = () => {
    const apiKey = 'bc50218d91157b1ba4f142ef7baaa6a0';
    const endpoint = `https://api.themoviedb.org/3/movie/${id}/reviews?api_key=${apiKey}`;

    fetch(endpoint, {
      method: 'GET',
    })
      .then((response) => response.json())
      .then((data) => {
        setData(data.results);
      })
      .catch((error) => {
        console.log(error);
      });
  };

  const list = data.slice(0, 2).map((review, index) => {
    const author = review.author;
    const content = review.content;

    return (
      <div key={`review-${index}`}>
        <h3>{author}</h3>
        <p>{content}</p>
      </div>
    );
  });

  return (
    <div className="movie__details__meta">
      <h2>Reviews</h2>
      {list.length > 0 ? list : <p>There are no available reviews.</p>}
    </div>
  );
};
