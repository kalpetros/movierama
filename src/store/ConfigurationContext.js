import React, { createContext, useState, useEffect } from 'react';

export const ConfigurationContext = createContext({});

export const ConfigurationStateProvider = ({ children }) => {
  const [state, setState] = useState({});

  useEffect(() => {
    fetchData();
  }, []);

  const fetchData = () => {
    const apiKey = 'bc50218d91157b1ba4f142ef7baaa6a0';
    const endpoint = `https://api.themoviedb.org/3/configuration?api_key=${apiKey}`;

    fetch(endpoint, {
      method: 'GET',
    })
      .then((response) => response.json())
      .then((data) => {
        setState(data.images);
      })
      .catch((error) => {
        console.log(error);
      });
  };

  return (
    <ConfigurationContext.Provider value={{ state }}>
      {children}
    </ConfigurationContext.Provider>
  );
};
