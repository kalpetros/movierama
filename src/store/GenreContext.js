import React, { createContext, useEffect, useState } from 'react';

export const GenreContext = createContext([]);

export const GenreStateProvider = ({ children }) => {
  const [state, setState] = useState([]);

  useEffect(() => {
    fetchData();
  }, []);

  const fetchData = () => {
    const apiKey = 'bc50218d91157b1ba4f142ef7baaa6a0';
    const endpoint = `https://api.themoviedb.org/3/genre/movie/list?api_key=${apiKey}`;

    fetch(endpoint, {
      method: 'GET',
    })
      .then((response) => {
        if (response.status !== 200) throw new Error(response.status);
        return response.json();
      })
      .then((data) => {
        setState(data.genres);
      })
      .catch((error) => {
        console.log(error);
      });
  };

  return (
    <GenreContext.Provider value={{ state }}>{children}</GenreContext.Provider>
  );
};
