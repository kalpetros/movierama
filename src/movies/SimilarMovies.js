import React, { useState, useEffect, useContext } from 'react';
import { ConfigurationContext } from '../store/ConfigurationContext';

export const SimilarMovies = (props) => {
  const [data, setData] = useState([]);
  const { state: configuration } = useContext(ConfigurationContext);
  const { id: id } = props;

  useEffect(() => {
    fetchSimilarMovies();
  }, []);

  const buildImageUrl = (path, size) => {
    const baseUrl = configuration.secure_base_url;
    const posterSize = configuration.poster_sizes[size];
    let imageUrl = null;

    if (path) {
      imageUrl = `${baseUrl}${posterSize}${path}`;
    }

    return imageUrl;
  };

  const fetchSimilarMovies = () => {
    const apiKey = 'bc50218d91157b1ba4f142ef7baaa6a0';
    const endpoint = `https://api.themoviedb.org/3/movie/${id}/similar?api_key=${apiKey}`;

    fetch(endpoint, {
      method: 'GET',
    })
      .then((response) => response.json())
      .then((data) => {
        setData(data.results);
      })
      .catch((error) => {
        console.log(error);
      });
  };

  const list = data.map((movie, index) => {
    const imageUrl = buildImageUrl(movie.poster_path, 2);

    return (
      <a key={`similar-movie-${index}`} className="movie__details__image">
        <img src={imageUrl} alt={movie.title} />
      </a>
    );
  });

  let content = <p>We couldn't find any similar movies.</p>;

  if (list.length > 0) {
    content = <div className="movie__details__list">{list}</div>;
  }

  return (
    <div className="movie__details__meta">
      <h2>Similar Movies</h2>
      {content}
    </div>
  );
};
