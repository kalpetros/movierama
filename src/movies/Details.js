import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Overview } from './Overview';
import { SimilarMovies } from './SimilarMovies';
import { Reviews } from './Reviews';

export const Details = (props) => {
  const [view, setView] = useState('overview');
  const { id: id, isSelected: isSelected } = props;

  const menuSelectedClass = 'movie__details__menu--selected';
  let baseClass = isSelected
    ? 'movie__details movie__details--active'
    : 'movie__details movie__details--inactive';
  let currentView = null;

  const handleChangeView = (e) => {
    const id = e.currentTarget.dataset.id;
    setView(id);
  };

  if (isSelected) {
    if (view === 'overview') {
      currentView = <Overview id={id} />;
    } else if (view === 'reviews') {
      currentView = <Reviews id={id} />;
    } else if (view === 'similar') {
      currentView = <SimilarMovies id={id} />;
    }
  }

  return (
    <div className={baseClass}>
      {currentView}
      <div className="movie__details__menu">
        <ul>
          <li
            data-id="overview"
            className={view === 'overview' ? menuSelectedClass : null}
            onClick={handleChangeView}
          >
            Overview
          </li>
          <li
            data-id="reviews"
            className={view === 'reviews' ? menuSelectedClass : null}
            onClick={handleChangeView}
          >
            Reviews
          </li>
          <li
            data-id="similar"
            className={view === 'similar' ? menuSelectedClass : null}
            onClick={handleChangeView}
          >
            Similar Movies
          </li>
        </ul>
      </div>
    </div>
  );
};

Details.defaultProps = {
  id: 1,
  isSelected: false,
};

Details.propTypes = {
  id: PropTypes.number.isRequired,
  isSelected: PropTypes.bool.isRequired,
};
