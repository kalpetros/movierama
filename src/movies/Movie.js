import React, { useContext, useState } from 'react';
import PropTypes from 'prop-types';

import { ConfigurationContext } from '../store/ConfigurationContext';
import { Details } from './Details';
import { GenreContext } from '../store/GenreContext';
import { buildImageUrl } from '../utils';

export const Movie = (props) => {
  const [isSelected, setIsSelected] = useState(false);
  const { state: configuration } = useContext(ConfigurationContext);
  const { state: genres } = useContext(GenreContext);
  const { data: data } = props;

  const handleClick = (e) => {
    setIsSelected((p) => (p ? false : true));
    const element = document.getElementById(`movie-${data.id}`);
    element.scrollIntoView({ behavior: 'smooth' });
  };

  let style = {};
  let overview = data.overview.length === 0 ? 'N/A' : data.overview;

  if (data.overview.length > 200) {
    overview = `${data.overview.substring(0, 200)}...`;
  }

  let legend = <p>{overview}</p>;
  const posterUrl = buildImageUrl(configuration, data.poster_path, 2);
  const year = new Date(data.release_date).getFullYear();

  let genresText = null;
  let genreList = data.genre_ids.map((id) => {
    const label = genres.find((g) => g.id === id);
    return label.name;
  });

  if (genreList.length > 0) {
    genresText = genreList.join(', ');
  }

  if (isSelected) {
    const popularity = Math.round(data.popularity, 0);
    const releaseDate = new Date(data.release_date);
    const date = releaseDate.getDate();
    const month = releaseDate.getMonth() + 1;
    const year = releaseDate.getFullYear();
    const dateString = `${date}/${month}/${year}`;
    const backgroundImage = buildImageUrl(configuration, data.poster_path, 5);

    legend = (
      <>
        <h3>Release date</h3>
        <p>{dateString}</p>
      </>
    );

    genresText = `Popularity - ${popularity}`;

    style = {
      backgroundImage: `url(${backgroundImage})`,
      backgroundRepeat: 'no-repeat',
      backgroundSize: 'cover',
    };
  }

  return (
    <div className="movie" style={style}>
      <div
        className="movie__wrapper"
        id={`movie-${data.id}`}
        onClick={handleClick}
      >
        <div className="movie__wrapper__content">
          <div className="movie__wrapper__image">
            <img src={posterUrl} alt="movie_poster" />
          </div>
          <div className="movie__wrapper__body">
            <div className="movie__wrapper__meta">
              <h2>
                {data.title} ({year})
              </h2>
              {legend}
            </div>
            <div className="movie__wrapper__footer">
              <div>
                <h3 className="font-bold">{genresText}</h3>
              </div>
              <div>
                <h3 className="font-bold text-right">
                  {data.vote_average} / 10
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
      <Details id={data.id} isSelected={isSelected} />
    </div>
  );
};

Movie.propTypes = {
  data: PropTypes.object.isRequired,
};
