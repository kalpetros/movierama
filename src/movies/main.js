import React, { useContext } from 'react';

import { MoviesContext } from '../store/movies/context';
import { Movie } from './Movie';

export const Movies = () => {
  const { movies } = useContext(MoviesContext);

  const list = movies.map((movie, index) => {
    return <Movie key={`movie-${index}`} data={movie} />;
  });

  return list;
};
