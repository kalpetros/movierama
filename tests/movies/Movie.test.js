import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

import { Movie } from '../../src/movies/Movie';
import { ConfigurationContext } from '../../src/store/ConfigurationContext';
import { GenreContext } from '../../src/store/GenreContext';

describe('<Movies />', () => {
  let wrapper;

  it('should match the snapshot', () => {
    const mockConfigurationContext = {
      state: {
        poster_sizes: ['w500', 'w780', 'original'],
        secure_base_url: 'https://image.tmdb.org/t/p/',
      },
    };
    const mockGenreContext = {
      state: [
        {
          id: 28,
          name: 'Action',
        },
      ],
    };
    const mockData = {
      genre_ids: [28],
      id: 619592,
      overview: 'Test overview',
      popularity: 50,
      poster_path: '/ucktgbaMSaETUDLUBp1ubGD6aNj.jpg',
      release_date: '2020-07-02',
      title: 'Test title',
      vote_average: 5,
    };

    wrapper = mount(
      <ConfigurationContext.Provider value={mockConfigurationContext}>
        <GenreContext.Provider value={mockGenreContext}>
          <Movie data={mockData} />
        </GenreContext.Provider>
      </ConfigurationContext.Provider>
    );

    expect(wrapper.html()).toMatchSnapshot();
  });
});
