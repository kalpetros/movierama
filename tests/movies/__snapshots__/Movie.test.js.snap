// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Movies /> should match the snapshot 1`] = `"<div class=\\"movie\\"><div class=\\"movie__wrapper\\" id=\\"movie-619592\\"><div class=\\"movie__wrapper__content\\"><div class=\\"movie__wrapper__image\\"><img src=\\"https://image.tmdb.org/t/p/original/ucktgbaMSaETUDLUBp1ubGD6aNj.jpg\\" alt=\\"movie_poster\\"></div><div class=\\"movie__wrapper__body\\"><div class=\\"movie__wrapper__meta\\"><h2>Test title (2020)</h2><p>Test overview</p></div><div class=\\"movie__wrapper__footer\\"><div><h3 class=\\"font-bold\\">Action</h3></div><div><h3 class=\\"font-bold text-right\\">5 / 10</h3></div></div></div></div></div><div class=\\"movie__details movie__details--inactive\\"><div class=\\"movie__details__menu\\"><ul><li data-id=\\"overview\\" class=\\"movie__details__menu--selected\\">Overview</li><li data-id=\\"reviews\\">Reviews</li><li data-id=\\"similar\\">Similar Movies</li></ul></div></div></div>"`;
